Unicam Media Controller support

6by9 opened the pull request:
Updates bcm2835-unicam to allow it to use Media Controller fully for the more complex pipelines.

~This does still want to be updated to add a DT switch between MC-centric and video-centric in addition to the module parameter~. There's then the discussion over which we adopt as the default.
@naushir has patches to make libcamera work with the new scheme, and that is the way we want to go, but it may need some coordination in changeover.

Commit:
media: v4l: subdev: Improve link format validation debug messages

Commit db8e94e7cf27d8bc101ef5b8ee5c1af77cd5b1c9 upstream

The existing link format validation failure debug message in media-entity.c
helped to pinpoint the point of failure but provided no additional
information what's wrong. Tell the user exactly why the validation failed.

Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>

6by9 committed:
media/bcm2835-unicam: Parse pad numbers correctly

The driver was making big assumptions about the source device
using pad 0 and 1, which doesn't follow for more complex
devices where Unicam's source device may be a sink device for
something else.

Read the pad numbers through media controller, and reference
them appropriately.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media/bcm2835-unicam: Add support for configuration via MC API

Adds Media Controller API support for more complex pipelines.
libcamera is about to switch to using this mechanism for configuring
sensors.

This can be enabled by either a module parameter, or device tree.

Various functions have been moved to group video-centric and
mc-centric functions together.

Based on a similar conversion done to ti-vpe.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

pelwell commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -2293,30 +2354,6 @@ static void unicam_return_buffers(struct unicam_node *node,
 	spin_unlock_irqrestore(&node->dma_queue_lock, flags);
 }
 
-static int unicam_check_format(struct unicam_node *node)
```
I think this hunk should be squashed with the previous commit - it was added there and not used.

6by9 replied:
> "Added a couple of extra patches that will be squashed into the "Add support for configuration via MC API", but Naush may have started reviewing."

There are further changes coming which backport work being done on mainline, but I wanted Naush to look at the basic changes first.

pelwell replied:
It would be kinder to squash it now than to have Naush review a new function which is never used.

pelwell commented on arch/arm64/configs/bcm2711_defconfig:
```diff
@@ -1404,6 +1405,7 @@ CONFIG_ANDROID=y
+CONFIG_MUX_GPIO=m
 CONFIG_EXT4_FS=y
 CONFIG_EXT4_FS_POSIX_ACL=y
 CONFIG_EXT4_FS_SECURITY=y
```
Some people out there use the arm64 bcmrpi3_defconfig, so let's keep it up-to-date.

6by9 replied:
OK, will do.

6by9 replied:
Dropped from the PR for now.

pelwell commented on arch/arm/boot/dts/overlays/imx219-mux-overlay.dts:
```diff
@@ -0,0 +1,235 @@
+		};
+	};
+
+	__overrides__ {
```
This isn't necessary, but I won't object if you'd prefer it to be retained as a placeholder.

6by9 replied:
Legacy. This overlay is very rough and ready (hence no README) to prove the concept of two image sensors off a mux such as https://thepihut.com/products/multi-camera-doubleplexer-stereo-module-v2. I'd started expecting to do 2 imx219's (hence the name), but ended up with an imx219 and an imx477.

What we really want is a way of specifying the sensor name attached to each port, and then import the relevant config from the existing overlays (imx219, imx477, ov5647, ov9281, ov7251, tc358743, imx517, or any others). I don't think that is really practical though, but I could be wrong.

The main thing for the PR is the driver changes, and we'll have to think about whether we want overlays supporting the mux as a further change.

6by9 replied:
imx219-mux-overlay removed from the PR.

pelwell commented on arch/arm/boot/dts/overlays/imx219-mux-overlay.dts:
```diff
@@ -0,0 +1,235 @@
+// SPDX-License-Identifier: GPL-2.0-only
```
The overlay looks fine, but it needs a brief description in the dreaded README.

naushir commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -2593,7 +2610,20 @@ static int unicam_probe_complete(struct unicam_device *unicam)
+				unicam->node[source_pads].src_pad_id = i;
+				unicam_err(unicam, "source pad %u is index %u\n",
+					   source_pads, i);
+			}
```
Change to unicam_dbg?

6by9 replied:
Done

naushir commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -2411,15 +2673,69 @@ static int register_node(struct unicam_device *unicam, struct unicam_node *node,
+
+static void unicam_mc_set_default_format(struct unicam_node *node, int pad_id)
+{
+	if (pad_id == IMAGE_PAD) {
```
I still wonder if this should query the subdev for the format, but it's not something other drivers do...

6by9 replied:
This conversion is based largely on Tomi's conversion of ti-vpe, and that doesn't do it.
AIUI The only interaction between the subdevs should be during link_validate, and we do that now.

naushir commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -2776,6 +3180,14 @@ static int unicam_probe(struct platform_device *pdev)
+	unicam->mc_api = media_controller;
+	if (of_property_read_bool(pdev->dev.of_node, "brcm,media-controller"))
+		unicam->mc_api = true;
+
```
Should we be updating imx477 and imx219 overlays to use this?

6by9 replied:
Yes, although with an override to disable it.
ov5647 as well, and presumably imx517. Basically anything that is expected to use libcamera should default to it being on, and other sensors could default to off (eg ov9281 and ov7251).

naushir replied:
Well, in that case, should it default to MC and have a switch to turn it off...?

6by9 replied:
Things like ADV728x and TC358743 remain with it off.

It's 3rd party overlays/drivers that I'm thinking of. They currently work, so making them need changes to keep them working is a bit mean.

naushir commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -1267,902 +1319,1216 @@ static int unicam_g_fmt_meta_cap(struct file *file, void *priv,
+	.vidioc_enum_fmt_meta_cap	= unicam_mc_enum_fmt_meta_cap,
+	.vidioc_g_fmt_meta_cap		= unicam_mc_g_fmt_meta_cap,
+	.vidioc_s_fmt_meta_cap		= unicam_g_fmt_meta_cap,
+	.vidioc_try_fmt_meta_cap	= unicam_g_fmt_meta_cap,
```
Typo!

6by9 replied:
Fixed

6by9 commented:
Rebased, fixed up comments, added updates to all CSI source overlays, and squashed.

Overlays shouldn't change now (expect based on review comments), but v4l2-compliance is complaining so I need to fix that up.
